source/models/users/Hunter.js

```diff
@@ -104,9 +113,50 @@ exports.Hunter = class Hunter extends Model {
 		return Math.min(slots, maxSimBounties);
 	}
 
-	// slotWorth(slotNum) {
-	// 	return Math.floor(6 + 0.5 * this.level - 3 * slotNum + 0.5 * slotNum % 2);
-	// }
+	slotWorth(slotNum) {
+		return Math.floor(6 + 0.5 * this.level - 3 * slotNum + 0.5 * slotNum % 2);
+	}
+
+	/**
+	 * @param {Guild} guild
+	 * @param {number} points
+	 * @param {boolean} ignoreMultiplier
+	 * @returns {string} level-up text
+	 */
+	async addXP(guild, points, ignoreMultiplier) {
+		const guildProfile = await database.models.Guild.findByPk(guild.id);
```

Comment:
Models having partial references to other models in a way that isn't that model owning an instance of the other is usually a pitfall that leads to circular references. Are we confident that we won't have Guild pulling for Hunters in the future?

Reply:
I have no certainty on that question. On the one hand, Guild level is derived from total Hunter level, but most uses of that doesn't live on the Guild entity? I mostly don't think we have enough pieces in play to make the call yet.

Reply:
I don't know how to put this especially lightly, but not knowing where data belongs like this is an incredibly bad sign. The most important factor of a relational database is its constructed and structured relationships, and data that doesn't have proper relational setup should either be treated as if it is everywhere all the time (bad because it makes it harder to work on code by expanding context to things that don't actually belong) or as if it doesn't exist always (since the only reason why data has value in this scheme is that it has a relationship to some other data and some concept/table encapsulating it). So, if this information is to exist right now, we want to make sure it fits in with a relationship as it is created in the codebase.

For a recommendation: this data does not belong on the individual hunter at all, since it relates to multiple database entries at once and not any individual hunter. The most appropriate place in my opinion is on the Guild, since that is the primary object that relates all the relevant data for this data point, in spite of (or really because of?) it being calculated using child components in the data hierarchy. We really don't want a new column for this, but we don't have to use one, either. (Maybe a better way of putting it: Guilds own the list of Hunters, so Guilds can know about Hunters and their data directly, while Hunters do not own and are not responsible for Guild data.) In my opinion, this seems like a good opportunity for a `virtual` field on the Guild that calculates the value when it is requested - this way we don't even need to do the calculations here at all. https://sequelize.org/docs/v6/core-concepts/getters-setters-virtuals/#virtual-fields

Reply:
(This is also a reason why most enterprise software has a DAO and/or Business Logic layer; it allows for disentangling the calculations from the data objects themselves.)

Reply:
Okay, to make the transition to having `guildXP` be virtual, I *think* this is all that you'll really need:
```
//...
guildXP: {
   type: DataTypes.VIRTUAL,
   async get() {
      return sequelize.query(`SELECT FLOOR(SQRT(SUM(level)/(:xpCoefficient + 1))) FROM hunters WHERE guildId = :guildId`, {
         replacements: [{ xpCoefficient: this.xpCoefficient }, { guildId: this.id }],
         type: QueryTypes.SELECT
      });
   }
}
//...
```

Reply:
Additional relevant docs:

- https://sequelize.org/docs/v6/core-concepts/raw-queries/
- https://sequelize.org/api/v6/class/src/sequelize.js~sequelize#instance-method-query